File: golang/3446_sort_matrix/sort_matrix.go
// Package main ...
package main

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func sortMatrix(grid [][]int) [][]int {
	n := len(grid)

	for i := 0; i < n; i++ {
		tmp := []int{}
		for j := 0; i+j < n; j++ {
			tmp = append(tmp, grid[i+j][j])
		}
		sort.Sort(sort.Reverse(sort.IntSlice(tmp)))
		for j := 0; i+j < n; j++ {
			grid[i+j][j] = tmp[j]
		}
	}

	for j := 1; j < n; j++ {
		tmp := []int{}
		for i := 0; j+i < n; i++ {
			tmp = append(tmp, grid[i][j+i])
		}
		sort.Ints(tmp)
		for i := 0; j+i < n; i++ {
			grid[i][j+i] = tmp[i]
		}
	}

	return grid
}

func main() {
	tests := []struct {
		grid   [][]int
		output [][]int
	}{
		{[][]int{{1, 7, 3}, {9, 8, 2}, {4, 5, 6}}, [][]int{{8, 2, 3}, {9, 6, 7}, {4, 5, 1}}},
		{[][]int{{0, 1}, {1, 2}}, [][]int{{2, 1}, {1, 0}}},
		{[][]int{{1}}, [][]int{{1}}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.output, sortMatrix(test.grid), index)
	}
}

File: rust/3446_sort_matrix/sort_matrix.rs
struct Solution;

impl Solution {
    pub fn sort_matrix(mut grid: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let n = grid.len();

        for i in 0..n {
            let mut tmp: Vec<i32> = (0..(n - i)).map(|j| grid[i + j][j]).collect();
            tmp.sort_by(|a, b| b.cmp(a));
            for j in 0..(n - i) {
                grid[i + j][j] = tmp[j];
            }
        }

        for j in 1..n {
            let mut tmp: Vec<i32> = (0..(n - j)).map(|i| grid[i][j + i]).collect();
            tmp.sort();
            for i in 0..(n - j) {
                grid[i][j + i] = tmp[i];
            }
        }

        grid
    }
}

fn main() {
    let tests = vec![
        (
            vec![vec![1, 7, 3], vec![9, 8, 2], vec![4, 5, 6]],
            vec![vec![8, 2, 3], vec![9, 6, 7], vec![4, 5, 1]],
        ),
        (vec![vec![0, 1], vec![1, 2]], vec![vec![2, 1], vec![1, 0]]),
        (vec![vec![1]], vec![vec![1]]),
    ];

    for (input, expected) in tests {
        assert_eq!(Solution::sort_matrix(input), expected);
    }
}
